Shared/Samples/Show device location history/ShowDeviceLocationHistoryView.swift

```diff
@@ -0,0 +1,234 @@
+// Copyright 2023 Esri
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import ArcGIS
+import SwiftUI
+
+struct ShowDeviceLocationHistoryView: View {
+    /// A Boolean value indicating whether to show an alert.
+    @State private var isShowingAlert = false
+    
+    /// The error shown in the alert.
+    @State private var error: Error? {
+        didSet { isShowingAlert = error != nil }
+    }
+    
+    /// A Boolean value indicating whether the tracking button is disabled.
+    @State private var trackingButtonIsDisabled = true
+    
+    /// The view model for this sample.
+    @StateObject private var model = Model()
+    
+    var body: some View {
+        MapView(map: model.map, graphicsOverlays: [model.trackOverlay, model.locationsOverlay])
+            .locationDisplay(model.locationDisplay)
+            .task {
+                do {
+                    try await model.startLocationDataSource()
+                    trackingButtonIsDisabled = false
+                } catch {
+                    // Shows an alert with an error if starting the data source fails.
+                    self.error = error
+                }
+            }
+            .onDisappear {
+                model.stopProcessingLocationChanges()
+                model.isTracking = false
+                model.stopLocationDataSource()
+            }
+            .toolbar {
+                ToolbarItem(placement: .bottomBar) {
+                    Button(model.isTracking ? "Stop Tracking" : "Start Tracking") {
+                        model.isTracking.toggle()
+                    }
+                    .disabled(trackingButtonIsDisabled)
+                }
+            }
+            .alert(isPresented: $isShowingAlert, presentingError: error)
+    }
+}
+
+private extension ShowDeviceLocationHistoryView {
+    /// The model used to store the geo model and other expensive objects
+    /// used in this view.
+    @MainActor
+    class Model: ObservableObject {
+        /// A Boolean value indicating whether to track the device location.
+        @Published var isTracking = false {
+            didSet {
+                if isTracking {
+                    startProcessingLocationChanges()
+                } else {
+                    stopProcessingLocationChanges()
+                }
+            }
+        }
```

Comment:
The model doesn't really use this property, so it doesn't make a lot of sense for it to be on the model. Instead, it could be a state property of the view itself:

```swift
@State private var isTracking = false
```

Then, instead of using a property observer to start/stop processing location changes, a task modifier could be used instead:

```swift
.task(id: isTracking) {
    guard isTracking else { return }
    await model.processingLocationChanges()
}

// In ShowDeviceLocationHistoryView.Model

func processingLocationChanges() async {
    for await location in locationDisplay.dataSource.locations {
        processLocationUpdate(lastPosition: location.position)
    }
}
```

That eliminates the need for `locationProcessingTask` and `stopProcessingLocationChanges()`. It also means that a toggle can be used:

```swift
Toggle(isTracking ? "Stop Tracking" : "Start Tracking", isOn: $isTracking)
    .toggleStyle(.button)
    .disabled(trackingButtonIsDisabled)
```

Reply:
Thank you. I like how the view's task modifier has the same lifetime as the view, so there is no need to manage a separate task by the user.